lib/swaps/Swaps.ts

```diff
@@ -141,218 +142,284 @@ class Swaps extends EventEmitter {
       } catch (err) {
-        this.logger.error(`Got exception from sendPaymentSync ${JSON.stringify(request.toObject())}`, err);
-        this.emit('swap.failed', deal);
+        this.logger.error('Got exception from sendPaymentSync: ' + ' ' + JSON.stringify(request.toObject()) + err.message);
+        this.setDealState(deal, SwapDealState.Error, err.message);
+        return 'Got exception from sendPaymentSync' + err.message;
       }
     } else {
-      // this is not a valid deal
-      this.logger.warn(`received swap request for unrecognized deal ${dealId}`);
-      // TODO: respond with a failure message, possibly penalize peer
+      // If we are here we are the taker
+      this.logger.debug('Executing taker code');
+      this.setDealStatus(deal, SwapDealStatus.AmountReceived);
+      return deal.r_preimage;
     }
 
-    const responsePacket = new packets.SwapResponsePacket(responseBody, requestPacket.header.id);
+  }
 
-    this.logger.debug('Sending swap response to peer: ' + JSON.stringify(responseBody));
+  private setDealState = (deal: SwapDeal, newState: SwapDealState, newStateReason: string): void => {
+    deal.state = newState;
+    deal.stateReason = newStateReason;
+    this.emit('swap.failed', deal);
+  }
 
-    peer.sendPacket(responsePacket);
+  private setDealStatus = (deal: SwapDeal, newStatus: SwapDealStatus): void => {
+    assert(deal.state === SwapDealState.Normal, 'deal state is Error. Can not change deal status');
+    switch (newStatus) {
+      case SwapDealStatus.SwapCreated:
+        assert(false, 'can not set deal status to SwapCreated.');
+        break;
+      case SwapDealStatus.SwapRequested:
+        assert(deal.myRole === SwapDealRole.Taker, 'SwapRequested can only be set by the taker');
+        assert(deal.status === SwapDealStatus.SwapCreated, 'SwapRequested can be only be set after SwapCreated');
+        break;
+      case SwapDealStatus.SwapAgreed:
+        assert(deal.myRole === SwapDealRole.Maker, 'SwapAgreed can only be set by the maker');
+        assert(deal.status === SwapDealStatus.SwapCreated, 'SwapAgreed can be only be set after SwapCreated');
+        break;
+      case SwapDealStatus.AmountSent:
+        assert(deal.myRole === SwapDealRole.Taker && deal.status === SwapDealStatus.SwapRequested ||
+          deal.myRole === SwapDealRole.Maker && deal.status === SwapDealStatus.SwapAgreed,
+            'AmountSent can only be set after SwapRequested (taker) or SwapAgreed (maker)');
+        break;
+      case SwapDealStatus.AmountReceived:
+        assert(deal.status === SwapDealStatus.AmountSent, 'AmountReceived can be only be set after AmountSent');
+        break;
+      case SwapDealStatus.SwapDone:
+        assert(deal.status === SwapDealStatus.AmountReceived, 'SwapDone can be only be set after AmountReceived');
+        break;
+      default:
+        assert(false, 'unknown deal status');
+    }
+
+    deal.status = newStatus;
+    switch (deal.status) {
```

Comment:
Why two switch statements? This logic can be combined with the switch statement above I believe without any change in behavior.

Reply:
Yes, it can be combined but I don't see what is wrong with two switch statements. 

The first switch is more for logic validation.

Do see a reason to combine these two switches.

Reply:
Because it adds extra comparisons and I think makes the code unnecessarily verbose and harder to follow - requiring us to look in two separate places to see what happens for each case. It's also just not a pattern I'm used to seeing anywhere.

Reply:
Ok. I merged the two so you will have it as you like and used to.

Now we have an ugly `if` after the assignment of the new `status` to check if `swap.paid` should be emitted.  This can't get into the switch since the `swap.paid` listeners may execute before the `deal .status` changed (after the switch).

I hope you like it like this better. I don't but don't consider this important at all.

Reply:
Just saw this. I don't think the `if` after the assignment is necessary. This method will finish executing before the listener executes.